helix-term/src/commands.rs

```diff
@@ -6509,6 +6511,215 @@ fn add_newline_impl(cx: &mut Context, open: Open) {
     doc.apply(&transaction, view.id);
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum MoveSelection {
+    Below,
+    Above,
+}
+
+#[derive(Clone)]
+struct ExtendedChange {
+    line_start: usize,
+    line_end: usize,
+    line_text: Option<Tendril>,
+    line_selection: Option<(usize, usize)>,
+}
+
+/// Move line or block of text in specified direction.
+/// The function respects single line, single selection, multiple lines using
+/// several cursors and multiple selections.
+fn move_selection(cx: &mut Context, direction: MoveSelection) {
+    let (view, doc) = current!(cx.editor);
+    let selection = doc.selection(view.id);
+    let text = doc.text();
+    let slice = text.slice(..);
+    let mut last_step_changes: Vec<ExtendedChange> = vec![];
+    let mut at_doc_edge = false;
+    let all_changes = selection.into_iter().map(|range| {
```

Comment:
Rather than iterating over all ranges here, `Selection::line_ranges` might be simpler. It's meant for line-wise stuff like this: it gives an iterator which merges overlapping and adjacent ranges. See its use in `:reset-diff-change` for an example